src/backends/gpu/stack.hpp

```diff
@@ -22,91 +25,85 @@ namespace gpu {
+    }
+
+    stack(unsigned capacity) {
+        base_ = allocator<base_type>().allocate(1);
+        base_->capacity = capacity;
+        base_->size = 0;
+        base_->data = allocator<value_type>().allocate(capacity);
     }
 
-    // Append a new value to the stack.
-    // The value will only be appended if do_push is true.
-    __device__
-    void push_back(const value_type& value) {
-        // Atomically increment the size_ counter. The atomicAdd returns
-        // the value of size_ before the increment, which is the location
-        // at which this thread can store value.
-        unsigned position = atomicAdd(&size_, 1u);
-
-        // It is possible that size_>capacity_. In this case, only capacity_
-        // entries are stored, and additional values are lost. The size_
-        // will contain the total number of attempts to push,
-        if (position<capacity_) {
-            data_[position] = value;
+    stack(): stack(0) {}
+
+    ~stack() {
+        if (base_) {
+            allocator<value_type>().deallocate(base_->data, base_->capacity);
+            allocator<base_type>().deallocate(base_, 1);
         }
     }
 
-    __host__
     void clear() {
-        size_ = 0;
+        base_->size = 0;
     }
 
     // The number of items that have been pushed back on the stack.
     // size may exceed capacity, which indicates that the caller attempted
     // to push back more values than there was space to store.
-    __host__ __device__
     unsigned size() const {
-        return size_;
+        return base_->size;
     }
 
     // The maximum number of items that can be stored in the stack.
-    __host__ __device__
     unsigned capacity() const {
-        return capacity_;
+        return base_->capacity;
+    }
+
+    base_type& base() {
+        return *base_;
     }
 
     value_type& operator[](unsigned i) {
-        EXPECTS(i<size_ && i<capacity_);
-        return data_[i];
+        EXPECTS(i<base_->size && i<base_->capacity);
+        return base_->data[i];
     }
 
     value_type& operator[](unsigned i) const {
-        EXPECTS(i<size_ && i<capacity_);
-        return data_[i];
+        EXPECTS(i<base_->size && i<base_->capacity);
+        return base_->data[i];
     }
 
     value_type* begin() {
-        return data_;
+        return base_->data;
     }
     const value_type* begin() const {
-        return data_;
+        return base_->data;
     }
 
     value_type* end() {
-        // Take care of the case where size_>capacity_.
-        return data_ + (size_>capacity_? capacity_: size_);
+        // Take care of the case where size>capacity.
+        return base_->data + std::min(base_->size, base_->capacity);
     }
     const value_type* end() const {
-        // Take care of the case where size_>capacity_.
-        return data_ + (size_>capacity_? capacity_: size_);
+        // Take care of the case where size>capacity.
+        return base_->data + std::min(base_->size, base_->capacity);
     }
```

Comment:
How about `size()` returning `std::min(size, capacity)`? the `size` member could be renamed to something more descriptive like `pushes`.

Reply:
I think this would be an improvement on the API. Do we care about the number of pushes other than to verify there hasn't been an overflow condition? If not, then I'd go with a boolean `overflow()` method instead, which would allow e.g. the wrap-safe test if that ever became a concern.

Reply:
I went ahead and did this (and implemented an `is_overflowed` method).